Implement Material MenuBar

gspencergoog opened the pull request:
## Description

This implements a `MenuBar` widget that can render a Material menu bar, and a `createMaterialMenu` function to create a cascading menu anywhere. The menus are drawn in the overlay, while the menu bar itself is in the regular widget tree. Keyboard traversal works between the two.

This implementation of the `MenuBar` uses the public global function `createMaterialMenu` to create a cascading menu that contains widgets representing the menu items. These menu items can be any kind of widget, but are typically `SubmenuButton`s that host submenus, or `MenuItemButton`s that have shortcut hints (but don't actually activate the shortcuts) and don't host submenus.

Cascading menus can be created outside of a `MenuBar` by calling `createMaterialMenu`. They can be either given a specific location to appear (global coordinate), or they can be located by a new `MenuAnchor` widget that wraps the control that opens them.

The developer may also create a `MenuController` to pass to the various menu primitives (`MenuBar` or `createMaterialMenu`) to associate menus so that they can be traversed together and closed together. Creating a controller is not required.

Things that are not implemented in this PR, but will need to be:
- Accelerators
- Showing the "selection trail" for selected menus. Parent menus should show as selected when a child is highlighted.
- Creating a horizontal menu bar that isn't a `MenuBar`. This is technically possible in the API, but `_MenuPanel` isn't public, and to write your own requires access to an internal type to set the orientation. Easy solution is to make `_MenuPanel` public, but I want to wait until there's a use case.
- A checkbox menu item that automatically shows and manages its own checkbox. You can technically do it right now with the prefix icon, but it's not very clean.
- Opening the menu initially with the keyboard (e.g. the Alt key on Linux/Windows, Ctrl-F2 on macOS).

## Related Issues
 - https://github.com/flutter/flutter/issues/23600

## Tests
 - Many tests for all of the menu operations and configurations

## Design Doc
- https://flutter.dev/go/cascading-menus

HansMuller commented on examples/api/lib/material/menu_bar/menu_bar.0.dart:
```diff
@@ -0,0 +1,212 @@
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    shortcutsEntry?.dispose();
+    // Register the shortcuts with the ShortcutRegistry so that they are available
```
This is a useful comment. Could also appear in the cascading menu example.

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+const bool _kDebugMenus = false;
+
+// How close to the edge of the safe area the menu will be placed.
+const double _kMenuViewPadding = 8.0;
```
These values come from the spec - https://m3.material.io/components/menus/specs (layout section). Should they be derived from design tokens?

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+/// register them. To be sure that selecting a menu item and triggering the
+/// shortcut do the same thing, it is recommended that they trigger the same
+/// [Intent] or call the same callback.
+/// {@endtemplate}
```
Should the link to the example also appear here, so that it's picked up wherever this template is used?

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+/// When a menu item with a submenu is clicked on, it toggles the visibility of
+/// the submenu. When the menu item is hovered over, the submenu will open, and
+/// hovering over other items will close the previous menu and open the newly
+/// hovered one. When those occur, [MenuButton.onOpen], and
```
those occur => those open/close transitions occur

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+  /// The Material elevation of the menu bar (if any).
+  ///
+  /// This is a [MaterialStateProperty], but [MenuBar] doesn't currently have
+  /// any states for it to respond to. Use [MaterialStatePropertyAll] to
```
You could choose to say that the menu bar is focused when any menu bar item is focused. Likewise selected when any menu bar menu is open.

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+  /// any states for it to respond to. Use [MaterialStatePropertyAll] to
+  /// initialize it.
+  ///
+  /// Defaults to the [MenuThemeData.barElevation] value of the ambient
```
MenuBar properties should have a concrete default, like elevation=0, that can be overridden by a widget (MenuBar) property or a MenuTheme property. The theme doesn't define defaults, it just enables the developers to override the defaults.

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+  /// An optional controller that allows outside control of the menu bar.
+  ///
+  /// You can use a controller to close any open menus from outside of the menu
+  /// bar using [MenuController.closeAll].
```
Should we remind the reader that if they provide a controller, they're responsible for disposing it?

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+  /// The preferred minimum height of the menu bar.
+  ///
+  /// Defaults to the value of [MenuThemeData.barMinimumHeight] if null.
+  final double? minimumHeight;
```
Do we need this or is it enough that the developer can enclose the MenuBar in a ConstrainedBox?

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+  /// any states for it to respond to. Use [MaterialStatePropertyAll] to
+  /// initialize it.
+  ///
+  /// Defaults to [MenuThemeData.barBackgroundColor] if null.
```
Here and elsewhere: we should reveal the actual defaults here or if that's impractical because the defaults are generated from tokens, then say that the default is defined by the Material spec.  The widget and theme properties don't define defaults they just enable the developer to override them.

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+                  child: Shortcuts(
+                    shortcuts: _kMenuTraversalShortcuts,
+                    child: _MenuPanel(
+                      elevation: (widget.elevation ?? menuTheme.barElevation ?? _TokenDefaultsM3(context).barElevation)
```
Default resolution like this is OK now because we're not resolving the MaterialStateProperties against any states. If we were then it would be necessary resolve each value:
```
widget.elevation?.resolve(states) ?? menuTheme.barElevation?.resolves(states) ?? _TokenDefaultsM3(context).barElevation!.resolve(states);
```
I'm only mentioning it because a future version of the code might introduce states and not get this right.

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+  }
+}
+
+/// An button for use in a [MenuBar] that can be activated by click or keyboard
```
An => A

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+  /// If neither [onSelected] nor [onSelectedIntent] are specified, then this
+  /// menu item is considered to be disabled.
+  ///
+  /// The default implementation returns null.
```
This sounds like the default onSelected callback is a function that returns null.

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+  /// button.
+  final ValueChanged<bool>? onHover;
+
+  /// Returns a callback, if any, to be invoked if the platform menu receives a
```
"Returns a callback, if any, to be invoked" could be A callback that is invoked

Same comment about the onSelectedIntent property

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+  /// The default implementation returns null.
+  final VoidCallback? onSelected;
+
+  /// Returns an intent, if any, to be invoked if the platform receives a
```
This sounds like a description of a platform-specific item. This may be a recurring problem...

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+  final ValueChanged<bool>? onHover;
+
+  /// Returns a callback, if any, to be invoked if the platform menu receives a
+  /// "Menu.selectedCallback" method call from the platform for this item.
```
This sounds like the description of a platform-specific menu bar callback.

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+  /// The focus node to use for the menu item button.
+  final FocusNode? focusNode;
+
+  /// An optional icon to display before the [label] label.
```
No text directionality dependency here. I suppose it doesn't bear saying.

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+
+  /// The background color for this [MenuItemButton].
+  ///
+  /// Defaults to the ambient [Theme]'s [ColorScheme.surface] if null.
```
NICE. That's the default!

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+  ///   that can be set instead of this property.
+  final MaterialStateProperty<Color?>? backgroundColor;
+
+  /// The foreground color for this [MenuItemButton].
```
I suppose it goes without saying that this applies to icons? Is it also used to create a DefaultTextStyle?

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+  ///    can be set instead of this property.
+  final EdgeInsetsDirectional? padding;
+
+  /// The text style for the text in this menu bar item.
```
Does foreground override the text style's color?

HansMuller reviewed (commented):
This all looks good; most of the nit-picking is to prove that I read the code.

Just a progress report, this is a big PR.

HansMuller commented on examples/api/lib/material/menu_bar/create_cascading_menu.0.dart:
```diff
@@ -0,0 +1,230 @@
+
+void main() => runApp(const MenuBarApp());
+
+enum MenuSelection {
```
I was curious what this all would look like if MenuSelection was an ordinary class that bundled the menu item's callback or submenu.  I just typed in all of the code that follows. Did not compile etc. Hopefully it's trivial enough to be self-explanatory.

A simple definition of MenuSelection as an ordinary class:

```dart
class MenuSelection {
  const class MenuSelection({ required this.label, this.shortcut, this.onSelected, this.menu });

  final String label;
  final MenuSerializableShortcut? shortcut;
  final VoidCallback onSelected
  final List<MenuSelection>? menu;

  static List<Widget> build(List<MenuSelection> selections) {
    Widget buildSelection(MenuSelection selection) {
      if (selection.menu != null) {
        return MenuButton(
          label: Text(selection.label),
          children: MenuSelection.build(selection.menu),
        );
      }
      return MenuItemButton(
        label: Text(selection.label),
        shortcut: selection.shortcut,
        onSelected: selection.onSelected,
      );
    }
    return selections.map<Widget>(buildSelection).toList();
  }

  static List<MenuSerializableShortcut> shortcuts(List<MenuSelection> selections) {
    List<MenuSerializableShortcut> value = MenuSerializableShortcut[];
    for (final MenuSelection selection in selections) {
      if (selection.menu != null) {
        value.addAll(MenuSelection.shorts(selection.menu));
      } else {
        if (selection.shortcut != null) {
          value.add(selection.shortcut);
        }
      }
    }
  }
}
``` 

The static methods would be used in init() and didUpdateWidget(). For example:
```dart
shortcutsEntry = ShortcutRegistry.of(context).addAll(MenuSelection.shortcuts(selections));
```
The calls to updateMenuEntry would be replaced by:
```dart
menuEntry.children = MenuSelection.build(selections);
```

And `selections` would be defined in the State subclass:
```dart
final List<MenuSelection> selections = <MenuSelection>[
  MenuSelection(
    label: 'About',
    onSelected: () {
      showAboutDialog(
        context: context,
        applicationName: 'MenuBar Sample',
        applicationVersion: '1.0.0',
      ),
    },
  ),
  MenuSelection(
    label: 'Show Message',
    shortcut: SingleActivator(LogicalKeyboardKey.keyS, control: true)
    onSelected: () {
      setState(() { showingMessage = !showingMessage; });
    },
  ),
  MenuSelection(
    label: 'Reset Message',
    shortcut: SingleActivator(LogicalKeyboardKey.escape),
    onSelected: !showingMessage ? null : () {
      setState() { showingMessage = false; };
    },
  ),
  MenuSelection(
    label: 'Hide Message'
    onSelected: () {
      setState(() { showingMessage = false; }),
    },
  ),
  MenuSelection(
    label: 'Background Color',
    menu: <MenuSelection>[
      MenuSelection(
        label: 'Red Background',
        shortcut: SingleActivator(LogicalKeyboardKey.keyR, control: true),
        onSelected: () { backgroundColor = Colors.red; },
      ),
      MenuSelection(
        label: 'Green Background',
        shortcut: SingleActivator(LogicalKeyboardKey.keyG, control: true),
        onSelected: () { backgroundColor = Colors.green; },
      ),
      MenuSelection(
        label: 'Blue Background',
        shortcut: SingleActivator(LogicalKeyboardKey.keyB, control: true),
        onSelected: () { backgroundColor = Colors.blue; },
      ),
    ),
  ),
];
```
This has the advantage of keeping the menu structure and the menus actions in one definition. The disadvantage is that MenuSelection is still a utility abstraction and in this version it hides the mapping from menu items to MenuItem and MenuItemButton.

gspencergoog replied:
If I understand this correctly, this is moving back in the direction of menus that are defined by non-widgets, and I thought we moved away from that precisely because they weren't widgets. I mean, now you're not prevented from making non-MenuItemButtons/MenuButtons, but you'd have to write your own equivalent to the `build` static method here to do it.

I'm getting the feeling that I must be missing something here, though.

Also, what's to stop someone from writing something like this on top of the code in this PR?

HansMuller replied:
Currently the example menu's implementation is spread across the app-specific MenuSelection enum class, the _activate() method, and the _updateMenuEntry method.  My goal was to reduce that by one: to a reusable MenuSelection class and the initializer for the selections state variable. The MenuSelection class could be copy-and-pasted and shared by all menus in an application. And yes, all of this can be built on top of the code in the PR, I didn't mean to suggest that it should be part of the new API.

Some mechanism that bridges between the shortcuts and the subset of  menu item onSelected callbacks that support those shortcuts is needed.  There are definitely other possibilities, for example the shared callbacks could be defined as methods within the State subclass and shared with the corresponding shortcut definitions.  I think it's fine to let idiomatic approaches to defining menus and shortcuts emerge as they see use in real apps.

gspencergoog replied:
Ahh, right, this is for the example, sorry, I missed that.  Sure, I'll see if I can consolidate in the way you suggest.

HansMuller replied:
Note also: this example is a special (common?) case where the menus are defined exclusively by MenuButtons and MenuItemButtons. It would probably be possible to write a utility function that registered the shortcuts by just walking the widget tree each time it was rebuilt.

gspencergoog replied:
OK, I modified the `menu_bar.0.dart` example to follow this model. I'm not sure it's "ideal" yet, but it gives another example of how to define menus. I left the other examples using enhanced enums so that they can see multiple forms.

gspencergoog commented on dev/manual_tests/lib/menu_bar.dart:
```diff
@@ -0,0 +1,577 @@
+    final ThemeData theme = Theme.of(context);
+    MenuThemeData menuTheme = MenuTheme.of(context);
+    if (_funkyTheme) {
+      menuTheme = menuTheme.copyWith(
```
OK, done.  I guess I'm just in the habit of using `copyWith` to be sure I don't do something weird with the theme.

gspencergoog commented on dev/manual_tests/lib/menu_bar.dart:
```diff
@@ -0,0 +1,577 @@
+            return Theme(
+              data: theme.copyWith(
+                visualDensity: _density,
+                menuTheme: _transparent
```
Done

gspencergoog commented on dev/manual_tests/lib/menu_bar.dart:
```diff
@@ -0,0 +1,577 @@
+                visualDensity: _density,
+                menuTheme: _transparent
+                    ? menuTheme.copyWith(
+                        barBackgroundColor: MaterialStatePropertyAll<Color?>(Colors.red.withOpacity(0.12)),
```
OK, I was just matching the declared types in the theme.

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+}
+
+class _MenuButtonState extends State<MenuButton> {
+  late _ChildMenuNode entry;
+  MenuEntry? childMenu;
```
Here and elsewhere: When reading this I found it confusing that `entry` is not a MenuEntry, but a MenuNode. Maybe unify terminology?

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+      );
+      childMenu = _createMenuEntryFromExistingNode(entry);
+    }
+    entry
```
if the menu was just created in the above if, this configuration work is happening twice... maybe use an else?

goderbauer reviewed (commented):
Comments from my initial read-through.

HansMuller commented on packages/flutter/lib/src/material/menu_theme.dart:
```diff
@@ -0,0 +1,354 @@
+///
+///  * [ThemeData], which describes the overall theme for the application.
+@immutable
+class MenuThemeData with Diagnosticable {
```
Developers will need to override the appearance of cascading menus, popup menus, and dropdown menus, and perhaps separately menu bars. It might not make sense to call anything just "menuTheme".

gspencergoog replied:
Okay.  Should I just make a plethora of themes, then?

HansMuller replied:
Themes for MenuBar and MenuButton sounds reasonable. We'll be adding a DropdownTheme (or maybe ComboBoxTheme if we find that we need a new name for the M3 version). There already is a PopupMenu theme.

HansMuller commented on packages/flutter/lib/src/material/menu_theme.dart:
```diff
@@ -0,0 +1,354 @@
+  /// The background color of a [MenuButton].
+  final MaterialStateProperty<Color?>? menuBackgroundColor;
+
+  /// The Material elevation of the [MenuButton].
```
"the MenuButton"?

HansMuller replied:
Just a grammar nit. It doesn't matter anymore :-)

HansMuller reviewed (commented):
Another update.

gspencergoog commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+    }
+  }
+
+  void focusButton() {}
```
Yeah, I just made it abstract and moved the empty implementation to _RootMenuNode, where it belongs.

gspencergoog commented on examples/api/test/material/menu_bar/create_cascading_menu.0_test.dart:
```diff
@@ -0,0 +1,103 @@
+
+    final State<example.MyCascadingMenu> menuState = tester.state(find.byType(example.MyCascadingMenu));
+
+    // ignore: avoid_dynamic_calls
```
removed.

gspencergoog commented:
Not yet ready to continue the review: still working on the ButtonStyleButton changes.

gspencergoog commented:
@HansMuller @goderbauer OK, I think it's safe to proceed with reviewing this some more.

I've converted the button classes to use ButtonStyleButton styles, and converted the `MenuButton` widget and `createMaterialMenu` function to take a new `MenuStyle` class that acts a lot like `ButtonStyle`, but for menus.  I've separated the themes into three themes (`MenuTheme`, `MenuBarTheme`, and `MenuButtonTheme`), and also into separate files.

The only thing I have left to do is to write copious tests for the new `MenuStyle` attributes, and their support in the menu classes. So, as long as you don't make it all the way down to the tests, I'm OK. :-)

Mainly I want to see if you want more large structural changes so I don't go down the test writing road too far if I need to rearrange still.

gspencergoog commented on packages/flutter/lib/src/material/theme_data.dart:
```diff
@@ -346,6 +348,10 @@ class ThemeData with Diagnosticable {
     FloatingActionButtonThemeData? floatingActionButtonTheme,
     IconButtonThemeData? iconButtonTheme,
     ListTileThemeData? listTileTheme,
+    MaterialBannerThemeData? bannerTheme,
```
To keep things alphabetized within each section as the comment above requests.

gspencergoog replied:
Oh, whoops, I sorted by type not name.

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3274 @@
+
+  /// The optional shortcut that selects this [MenuItemButton].
+  ///
+  /// This shortcut is only enabled when [onPressed] is set.
```
Do we need to include the boilerplate about registering shortcuts here?

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3274 @@
+    // Since we don't want to use the theme style or default style from the
+    // TextButton, we merge the styles, merging them in the right order when
+    // each type of style exists.
+    final ButtonStyle mergedStyle =
```
This multiway merge is pretty slick. It might be a little more efficient to compute `widget.defaultStyleOf(context)` just once.

gspencergoog replied:
That's not necessary.  If the component before it is null, then it won't be executed anyhow.  It only gets executed once.

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3274 @@
+  }
+}
+
+/// A menu item widget that displays a hierarchical cascading menu as part of a
```
MenuButton can be used outside of menu bars

HansMuller commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3274 @@
+/// Creates a new cascading menu given the focus node for the controlling
+/// widget.
+///
+/// Calling `createMaterialMenu` creates a new cascading menu controlled by
```
How does this work if the menu is to be shown when the user clicks right somewhere on the app? In that case the default origin of the menu should be aligned with the origin of the mouse event,

gspencergoog replied:
Currently, that isn't possible.  I need a `BuildContext` in order to copy the themes into the overlay, and a single point doesn't give me that context, so I required a `GlobalKey` (I used to require a `FocusNode`, which also has a `BuildContext`, but I decided that was too restrictive).  I could make a different function that takes a `Rect` and a `BuildContext`, but the problem then is that the build context isn't guaranteed to be valid next frame, so I'm back to taking a `GlobalKey` which will update itself. I could add an optional `Rect` to use instead of the rect from the `GlobalKey`'s context, however, so maybe that's the solution, but you'd still need to supply a GlobalKey that lets me know what context I should copy the themes from. Once we have the `OverlayPortal` in, that could simplify this a lot.

HansMuller replied:
Maybe the controller's open method should have an optional  parameter that's the global coordinates of the menu's origin (subject to adjustment of course).

gspencergoog replied:
There already is an `alignmentOffset` member on `MenuEntry` that you can update, and it will update the location of the menu automatically. It is in relative coordinates from the alignment position, so if you specify `Alignment.topLeft` as the alignment, then it is in local coordinates of the rect that the `GlobalKey`'s context describes.

I could add an optional `Offset` to the "open" that would take a global coordinate, and then translate that to a local coordinate internally and set the `alignmentOffset` from that so that it would be internally consistent.

gspencergoog replied:
Actually, resolving the correct local `alignmentOffset` is actually really hard at the point where we're opening the menu, so I think instead I'll just have an attribute that is normally null, but if you set it, overrides the `alignment`/`alignmentOffset` settings and just puts it at that global location.

HansMuller replied:
That seems reasonable to me. A context or background menu that pops up under the mouse on right click is a really common desktop UI idiom, so an example or a feature of an existing example that demo'd as much would be helpful.

HansMuller reviewed (commented):
Progress report. Mostly this looks great. I'm concerned about how one would defined the origin context menus.

gspencergoog commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3274 @@
+///
+/// This widget represents a leaf entry in a menu that is typically part of a
+/// [MenuBar], but may be used independently, or as part of a menu created with
+/// [createMaterialMenu].
```
Done.

gspencergoog commented:
OK, I think I have added all the tests I'm going to add for now.

bleroux commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3396 @@
+///   work with the [Actions] and [Shortcuts] system.
+/// * [CallbackShortcuts] to define shortcuts that simply call a callback and
+///   don't involve using [Actions].
+/// * [PlatformMenuBar], a class that renders similar menu bar items from a
```
`PlatformMenuBar` is already listed above (line 322)

gspencergoog replied:
Thanks!

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3396 @@
+    // Listeners need to always be notified in the next frame because the major
+    // listener for this object is in the overlay, and it can't rebuild during
+    // the build that triggers the change.
+    SchedulerBinding.instance.addPostFrameCallback((Duration _) {
```
guard against only scheduling one callback?

gspencergoog replied:
Sure, done.

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3396 @@
+  _RootMenuNode get root => this;
+
+  @override
+  _MenuNode get parent => throw UnimplementedError('Tried to get the parent of the root node');
```
Probably more of a UnsupportedError maybe?

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3396 @@
+  FocusScopeNode menuScopeNode;
+
+  @override
+  void focusButton() {}
```
nit: leave a comment why this is empty?

gspencergoog replied:
Done.

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3396 @@
+  set alignmentOffset(Offset value) {
+    if (_alignmentOffset != value) {
+      _alignmentOffset = value;
+      _notifyNextFrame();
```
Does that mean that any change to the menu will be delayed by one frame? Why is that necessary? If I change the alignment in a button handler, it should become active the very next frame? Not two frames later?

gspencergoog replied:
Yes, it will be delayed by a frame.  The menu bar or the widget that you've attached the menu to is in the main widget tree.  The menu is in the overlay, so marking it dirty needs to happen between frames.

goderbauer replied:
What if I am already between frames though (e.g. in an onpressed handler of a button) when reconfiguring the Node? In that scenario, updates will now be delayed by two frames, I think. Timeline:

(we start being between to frames)
* Node is reconfigured, _notifyNextFrame schedules a postframe callback (P) and requests a frame (F1)
* Scheduled frame (F1) happens
* At the end of (F1), (P) executes and listeners of the node are informed
* Listeners dirty the tree to adapt to the updated node, which schedules another frame F2
* F2 is produced to show the updated node configuration on screen.

THe update is now two frames late. We could have applied it in F1, though.

I think you just need to check in _notifyNextFrame whether we are between frames and if we are notifyLIsteners directly, if we are inside a frame, do the postframecallback thing.

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3396 @@
+}
+
+bool _menuDebug(String message, [Iterable<String>? details]) {
+  if (_kDebugMenus) {
+    debugPrint('MENU: $message');
+    if (details != null && details.isNotEmpty) {
+      for (final String detail in details) {
+        debugPrint('    $detail');
+      }
+    }
+  }
```
wrap this in an assert so it is guaranteed to get removed in release builds?

gspencergoog replied:
Each usage is already wrapped, but for insurance, sure.

(I could just assert we're not in release mode, right? LOL! (j/k))

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3396 @@
+  String get debugDescription => 'Menu_MouseCursor';
+}
+
+bool _menuDebug(String message, [Iterable<String>? details]) {
```
to follow naming conventions, start this one with "_debug"?

gspencergoog replied:
Done.

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,2896 @@
+  bool get enabled => true;
+  FocusScopeNode get menuScopeNode;
+
+  List<_ChildMenuNode> children = <_ChildMenuNode>[];
```
Everything in this class is public, though. It doesn't have a private API.

And is it public API to set children to a completely different List or should this be final?

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3396 @@
+
+// Base class for all menu nodes that make up the menu tree, to allow walking of
+// the tree for navigation.
+abstract class _MenuNode with DiagnosticableTreeMixin, ChangeNotifier {
```
The _MenuNode / MenuEntry split remains a bit suspect to me. It feels like these should bit one and the same thing with a public API to be used by anyone and a private API that's only used within this file.

After all, MenuController is also just one class and not split into two. It has a public API (e.g. `closeAll`) and a file-private API (e.g. `_menuOpened`)

gspencergoog replied:
Yeah, OK.  I'm sure I can merge them.  I'll work on that.

gspencergoog commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3396 @@
+  ///
+  /// The controller itself can be listened to for state changes (it is a
+  /// [ChangeNotifier]).
+  static MenuController of(BuildContext context) {
```
Good idea. Done.

gspencergoog commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3396 @@
+    return context.dependOnInheritedWidgetOfExactType<_MenuControllerMarker>()?.controller;
+  }
+
+  // Called by the _MenuEntry.open to notify the controller when a menu item has been
```
No.  It's supposed to be `_MenuNode`. I fixed all these.

gspencergoog commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3396 @@
+        onTapOutside: (PointerDownEvent event) {
+          MenuController.of(context).closeAll();
+        },
+        child: SingleChildScrollView(
```
Yes, I want to fix it too! This just *can't* be the only solution.

The issue is that none of the things I've tried properly wraps the contents without causing overflows.  I've tried all the combinations I can think of, and all of them seem to fail one way or another.

Let's talk about this too.

gspencergoog commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3396 @@
+          Padding(
+            padding: EdgeInsetsDirectional.only(start: horizontalPadding),
+            child: const Icon(
+              Icons.arrow_right, // Automatically switches with text direction.
```
Yeah, me too. Probably `Icons.arrow_that_automatically_switches_direction` was taken. :-)

gspencergoog commented on packages/flutter/lib/src/material/menu_button_theme.dart:
```diff
@@ -0,0 +1,136 @@
+/// overall [Theme]'s [ThemeData.menuTheme].
+///
+/// The [style]'s properties override [MenuItemButton]'s and [MenuButton]'s
+/// default style, i.e.  the [ButtonStyle] returned by
```
Done.

gspencergoog reviewed (commented):
OK, I've made all the changes except:

1) I haven't merged _MenuNode and MenuEntry.
2) I haven't replaced the non-scrolling scroll child with something better.
3) I haven't replaced the GlobalKey requirement with something better.

We should have a chat about the last two. I'll put something on the calendar.
I think I can do the first one, it'll just take a bit of work. 

Highlights:
- Removed all trace of `MenuItemGroup`
- Added/updated a bunch of words.
- Guarded scheduled post frame callbacks so they only schedule once.
- Made sure that the MenuController I create in `createMaterialMenu` is disposed.

and lots more.

bleroux commented on examples/api/test/material/menu_bar/create_material_menu.0_test.dart:
```diff
@@ -0,0 +1,96 @@
+    expect(find.text(example.kMessage), findsOneWidget);
+    expect(find.text('Last Selected: ${example.MenuSelection.showMessage.label}'), findsOneWidget);
+  });
+  testWidgets('Shortcuts work', (WidgetTester tester) async {
```
nit: missing line break

bleroux commented on packages/flutter/test/material/menu_bar_test.dart:
```diff
@@ -0,0 +1,1826 @@
+        tester.getRect(find.text(TestMenu.subMenu10.label)),
+        equals(const Rect.fromLTRB(534.0, 69.0, 688.0, 83.0)),
+      );
+      expect(
+          tester.getRect(find.ancestor(of: find.text(TestMenu.subMenu10.label), matching: find.byType(Material)).at(1)),
+          equals(const Rect.fromLTRB(466.0, 48.0, 696.0, 200.0)));
```
nit: extra indentation and missing comma before ending paranthesis
```suggestion
      expect(
        tester.getRect(find.ancestor(of: find.text(TestMenu.subMenu10.label), matching: find.byType(Material)).at(1)),
        equals(const Rect.fromLTRB(466.0, 48.0, 696.0, 200.0)),
      );
```

bleroux commented on packages/flutter/test/material/menu_style_test.dart:
```diff
@@ -0,0 +1,448 @@
+        tester.getRect(find.text(TestMenu.subMenu10.label)),
+        equals(const Rect.fromLTRB(366.0, 64.0, 520.0, 78.0)),
+      );
+      expect(
+          tester.getRect(find.ancestor(of: find.text(TestMenu.subMenu10.label), matching: find.byType(Material)).at(1)),
+          equals(const Rect.fromLTRB(350.0, 48.0, 602.0, 178.0)));
```
nit: same, extra indentation and missing comma before ending paranthesis

bleroux commented on packages/flutter/test/material/menu_bar_theme_test.dart:
```diff
@@ -0,0 +1,324 @@
+          menuChildren: <Widget>[
+            MenuItemButton(
+              onPressed: onPressed != null ? () => onPressed(TestMenu.subSubMenu100) : null,
+              shortcut: shortcuts[TestMenu.subSubMenu100],
```
I'm not sure it is worth it, consider renaming `subSubMenu100` (and similar) to `subSubMenu110`

bleroux commented on packages/flutter/test/widgets/tap_region_test.dart:
```diff
@@ -59,48 +59,48 @@ void main() {
-    expect(clickedOutside, isEmpty);
+    expect(tappedOutside, isEmpty);
   });
   testWidgets('TapRegionSurface detects inside taps', (WidgetTester tester) async {
```
nit: missing line break

bleroux commented on packages/flutter/test/widgets/tap_region_test.dart:
```diff
@@ -150,39 +150,139 @@ void main() {
-    expect(clickedInside, isEmpty);
+    expect(tappedInside, isEmpty);
+  });
+  testWidgets('Setting the group updates the registration', (WidgetTester tester) async {
```
nit: missing line break

bleroux commented on packages/flutter/test/widgets/tap_region_test.dart:
```diff
@@ -59,48 +59,48 @@ void main() {
     await click(find.text('No Group'));
     expect(
-        clickedOutside,
+        tappedOutside,
         unorderedEquals(<String>{
           'Group 1 A',
           'Group 1 B',
         }));
```
nit: extra indentation and missing comma before ending paranthesis
(here and in several places below)

bleroux reviewed (commented):
Very very interesting PR!

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3396 @@
+  }
+
+  RelativeRect _getMenuButtonRect() {
+    final RenderBox button = widget.node.context.findRenderObject()! as RenderBox;
```
It is possible that the configuration on the node stays unchanged while the button widget (aka its associated RenderObject) is moved to a new position on screen, though.

goderbauer commented on dev/manual_tests/lib/menu_bar.dart:
```diff
@@ -0,0 +1,614 @@
+            LogicalKeyboardKey.keyB,
+            control: true,
+          ),
+          onPressed: () {},
```
Yeah, when I first read through this I incorrectly thought this was supposed to be an example we showcase to devs. Only later did I realize that this is a manual test. Sorry about that.

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3396 @@
+    assert(debugCheckHasOverlay(context));
+    return AnimatedBuilder(
+      animation: _controller,
+      builder: (BuildContext context, Widget? ignoredChild) {
```
_MenuControllerMarker only needs to now when the controller is replaced with a new instance (in which case the entire build method needs to re-run, not just the builder), not when it notifies its listeners. So you should be able to move that out of the AnimatedBuilder as well...

HansMuller commented on packages/flutter/lib/src/rendering/shifted_box.dart:
```diff
@@ -656,9 +656,9 @@ class RenderConstrainedOverflowBox extends RenderAligningShiftedBox {
-/// A [RenderBox] that applies an arbitrary transform to its [constraints]
-/// before sizing its child using the new constraints, treating any overflow as
-/// error.
+/// A [RenderBox]  that applies an arbitrary transform to its constraints,
```
looks like an extra space snuck in here.

gspencergoog replied:
1) already did, in https://github.com/flutter/flutter/pull/110393
2) already fixed.

HansMuller reviewed (commented):
Maybe land the changes to ShiftedBox as a separate PR?

HansMuller commented on packages/flutter/lib/src/material/menu_style.dart:
```diff
@@ -0,0 +1,447 @@
+  /// The surface tint color of the menu's [Material].
+  ///
+  /// See [Material.surfaceTintColor] for more details.
+  final MaterialStateProperty<Color?>? surfaceTintColor;
```
CC @darrenaustin 
Presumably the API doc for Material.surfaceTintColor will explain that using Colors.transparent means "no surface tint".

darrenaustin replied:
Yup. I am prepping a PR with this support as we type.

HansMuller commented on packages/flutter/lib/src/material/menu_style.dart:
```diff
@@ -0,0 +1,447 @@
+  /// If there isn't sufficient space to open the menu with the given alignment,
+  /// and there's space on the other side of the button, then the alignment is
+  /// swapped to it's opposite (1 becomes -1, etc.), and the menu will try to
+  /// appear on the other side of the menu. If there isn't enough space there
```
```suggestion
  /// appear on the other side of the button. If there isn't enough space there
```

HansMuller commented on packages/flutter/lib/src/material/menu_style.dart:
```diff
@@ -0,0 +1,447 @@
+  /// right corner of the [MenuButton] that owns it. This applies for vertical
+  /// menus (e.g. created by a [MenuButton] or [createMaterialMenu]).
+  ///
+  /// For a [MenuBar], the default is [AlignmentDirectional.bottomStart].
```
We usually only document defaults in the component classes.

gspencergoog replied:
Okay, I removed that bit.

HansMuller commented on packages/flutter/lib/src/material/menu_style.dart:
````diff
@@ -0,0 +1,447 @@
+  /// ```dart
+  /// MenuStyle newStyle = style.copyWithout(backgroundColor: true);
+  /// ```
+  MenuStyle copyWithout({
````
For the sake of consistency we should a copyWithout method to ButtonStyle (in a separate PR).

gspencergoog committed:
Implement Material MenuBar and createCascadingMenu.

gspencergoog committed:
Some comment fixes, broken directional traversal

gspencergoog committed:
Fix traversal

gspencergoog committed:
Fix tests

gspencergoog committed:
Fix doc problems and analyzer fail.

gspencergoog committed:
Simplify anchor build

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3167 @@
+///
+/// The [dispose] method must be called when the menu handle is no longer
+/// needed. The object shouldn't be used after [dispose] is called.
+///
```
As commented on earlier: The distinction between MenuHandle and MenuController is not super obvious. The handle kinda feels like a controller as well. If we can't find a better solution for this, we may want to add some docs that contrast the two with each other here and in the MenuController doc.

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3167 @@
+  /// Call this when the menu should be shown to the user.
+  ///
+  /// The optional `position` argument will update the global coordinate where
+  /// the menu should be opened, taking into account the [_alignmentOffset].
```
nit: here and in the paragraph below: docs shouldn't refer to private variables.

gspencergoog replied:
Oh!  No, they shouldn't!

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3167 @@
+  /// If `position` is not given, and no `globalMenuPosition` was given to
+  /// [createMaterialMenu], the menu appears at the location that is
+  /// [_alignmentOffset] away from the alignment origin specified by
+  /// [MenuStyle.alignment] for the menu.
```
I thought the LayerLink or MenuAnchor from context is used for positioning if no explicit position is given?

gspencergoog replied:
Yes, that's correct.  I clarified the docs.

goderbauer reviewed (commented):
This is looking pretty good.

One thing that isn't clear to me (I probably just missed it in the 8000+ lines 😝): Why do we still need the button rect to position the menu? I thought the LayerLink stuff would take care of most of that...

gspencergoog committed:
Review Changes

gspencergoog reviewed (commented):
Still need to separate out the unrelated `MaterialStateProperty` changes.

gspencergoog commented on examples/api/lib/material/menu_bar/create_material_menu.0.dart:
```diff
@@ -0,0 +1,238 @@
+
+  @override
+  Widget build(BuildContext context) {
+    _updateMenu();
```
Actually, that does make it more complex: If the state object is a `ChangeNotifier`, then it can't notify until the next frame, since then it could be notifying a widget that was drawn in the overlay.

goderbauer replied:
Hm, I see. However, in the current implementation, if `MyCascadingMenu` (the widget that owns the menu handle) rebuilds for unrelated reasons (maybe because one of its dependencies changed, etc) that would close the menu because we dispose the handle on build, no? To the user this would look like an annoying glitch when the menu closes for reasons not obvious to them? In theory, the `build` method could be called every frame (that it isn't is just an optimization in the framework)...

Also: If a widget owning a menu happens to get rebuild frequently for reasons unrelated to the menu, we are doing a bunch of extra work in every build which isn't necessary if the menu isn't shown. Since menus are rather temporary, maybe they should only do this work if the menu is actually shown?

gspencergoog replied:
Perhaps I should just modify the example to not show updated state in the menu in response to a state change here?

goderbauer replied:
Since this is the introductory example, that would be fair to simplify it like that.

Still leaves the question of how you would do this in a real app if you need that? Maybe the answer is you'd just use package:provider or whatever to inject the data you want into the menu and package:provifder would take care of rebuilding locally?

gspencergoog replied:
Yeah, I think that's basically the answer.  I mean, I could do it here too using a `ChangeNotifier` and scheduling the update post-frame, but that seems like overkill for an example.

In the future, when we have `OverlayPortal`, I think this will get a lot easier, since it would just listen to a change notifier as normal, and wouldn't need to delay a frame.

gspencergoog replied:
Also, this is using `createMaterialMenu`, which is a slightly lower level API. If they use `MenuBar`, then the updating is taken care of for them.

gspencergoog committed:
Revert some unrelated changes

gspencergoog committed:
Try out renaming MenuButton to SubmenuButton

gspencergoog committed:
Fix some minor things with positioning and menu handles

gspencergoog committed:
Fix size of menu items

gspencergoog commented:
> One thing that isn't clear to me (I probably just missed it in the 8000+ lines 😝): Why do we still need the button rect to position the menu? I thought the LayerLink stuff would take care of most of that...

We still need the button rect because the custom layout needs to be able to position the menu by looking at the global coordinates of the menu button: it needs to know if the menu is too close to the edge of the view or not. The `LayerLink` takes care of when the view is resized or scrolled after it is positioned, but it doesn't know if the menu is near an edge or not.  If you have a menu up and scroll and get the menu too close to an edge, it won't relocate the menu (it'll just scroll off the edge), but that's an edge case that we probably don't need to worry about.

gspencergoog committed:
fix test

gspencergoog committed:
Fix doc comment

goderbauer commented on examples/api/lib/material/menu_bar/create_material_menu.0.dart:
```diff
@@ -0,0 +1,238 @@
+              focusNode: _buttonFocusNode,
+              onPressed: () {
+                if (_menuHandle.isOpen) {
+                  _menuHandle.close();
```
> MenuHandles are for individual menus. MenuControllers allow you to link otherwise unrelated menus so that you can close them all at once and tab between them when they are open.

I think it would be enough to just document this in the doc comments for MenuHandle and MenuController.

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3167 @@
+/// standalone widget, or as part of a menu defined by [createMaterialMenu].
+///
+/// This widget represents an item in a [MenuBar] or menu that has a submenu.
+/// Like the leaf [MenuItemButton], it shows a label with an optional leading or
```
I like `SubmenuButton` (or `SubMenuButton` - whatever is more correct english). MenuItemButton still feels ambiguous because isn't a SubMenuButton also a MenuItem? But if one of them has a less ambiguous name that's still a big win.

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3167 @@
+  MenuController? _internalMenuController;
+  FocusNode get _buttonFocusNode => widget.focusNode ?? _internalFocusNode!;
+  bool _waitingToFocusMenu = false;
+  final GlobalKey _buttonKey = GlobalKey();
```
Oh, I see. Makes sense. 

I noticed below that depending on how the menu is opened, different `context`s are used. Sometimes its the Builder context, sometimes its the GlobalKey context. Maybe we can unify that since I believe there is no meaning in those differences?

goderbauer replied:
I believe this little nit is still unaddressed:

> I noticed below that depending on how the menu is opened, different contexts are used. Sometimes its the Builder context, sometimes its the GlobalKey context. Maybe we can unify that since I believe there is no meaning in those differences?

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3167 @@
+
+    Widget child = CustomSingleChildLayout(
+      delegate: _MenuLayout(
+        buttonRect: anchor != null ? _getMenuButtonRect(anchor.anchorKey.currentContext!) : null,
```
Can you document this as a code comment somewhere? Future readers will probably appreciate the hint about what the rect and what the CompositedTransformTarget / LayerLink is used for.

goderbauer commented:
> We still need the button rect because the custom layout needs to be able to position the menu by looking at the global coordinates of the menu button: it needs to know if the menu is too close to the edge of the view or not. The LayerLink takes care of when the view is resized or scrolled after it is positioned, but it doesn't know if the menu is near an edge or not. If you have a menu up and scroll and get the menu too close to an edge, it won't relocate the menu (it'll just scroll off the edge), but that's an edge case that we probably don't need to worry about.

Can you document this (and the limitation you describe) in the code somewhere? This may be useful for a future reader.

gspencergoog committed:
goderbauer review changes

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3212 @@
+///   work with the [Actions] and [Shortcuts] system.
+/// * [CallbackShortcuts] to define shortcuts that simply call a callback and
+///   don't involve using [Actions].
+class MenuBar extends StatefulWidget with DiagnosticableTreeMixin {
```
What do you need the `DiagnosticableTreeMixin` for? I am a little afraid that this introduces bugs similar to what we saw with the TapRegion.

You can override `debugDescribeChildren` and `debugFillProperties` without adding this mixin.

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3212 @@
+  /// [createMaterialMenu], the menu will appear at the location that is
+  /// `alignmentOffset` away from the alignment origin specified by the
+  /// combination of the [MenuStyle.alignment] for the menu and the [MenuAnchor]
+  /// for the menu.
```
Maybe document that this MenuAnchor is looked up via the provided context.

goderbauer replied:
Maybe document in general what the context is for?

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3212 @@
+  /// for the menu.
+  void open(BuildContext context, {Offset? position}) {
+    assert(ChangeNotifier.debugAssertNotDisposed(this));
+    if (isOpen && position == _globalMenuPosition) {
```
what if it is given a different context - which could result in a different position?

goderbauer replied:
Maybe it would be easier to define that an open menu cannot be opend again?

goderbauer commented on packages/flutter/lib/src/material/menu_bar.dart:
```diff
@@ -0,0 +1,3212 @@
+      if (anchor == null || _globalMenuPosition != null) {
+        return true;
+      }
+      final _MenuHandleMarker? marker = anchor.anchorKey.currentContext!.findAncestorWidgetOfExactType<_MenuHandleMarker>();
```
Isn't _MenuHandleMarker an inheritedwidget that could/should be looked up via one of the `getFoo` methods instead of `findFoo`?

goderbauer reviewed (approved):
LGTM

goderbauer commented:
Just as I stamped this LGTM I actually had one more idea that may improve the API and may make it easier to integrate this with OverlayPortal and/or multi window in the future. I think it would do away with MenuHandle and createMaterialMenu altogether and would avoid passing contexts around. Rough API sketch:

```dart
final MenuController _controller = MenuController();

@override
Widget build(BuildContext context) {
  return MenuAnchor(
    controller: controller,
    // ... other config options that would have been passed to createMaterialMenu before
    menuBuilder: (BuildContext context) {
      return <Widget>[
        MenuItemButton(...),
        SubMenuButton(...),
        // ...
      ],
    },
    child: TextButton(
     onPressed: () {
       if (!controller.isOpen) {
         // maybe open needs to take an optional offset param to allow specifying where within the MenuAnchor's bounding box the menu should be shown...
         controller.open();
       } else {
         controller.close();
       }
     },
     child: Text('Toggle menu'),
    )
  );
}
```

Basically, the menu is always anchored at the MenuAnchor. It's context is used for positioning (so you want to wrap it tightly e.g. around a button or whatever UI element is supposed to be aligned with the menu). Now the widgets that make up the menu entries actually look like they are part of the widget tree and when we move menues over to OverlayPortal or multi window it is immediately obvious that they can access all inherited widgets above the MenuAnchor, because that's where they are anchored. 

Also, I think this will make it easier to move to OverlayPortal later (and multi window as well) because for both you kinda need an anchor in the widget tree.

It also simplifies the API by removing concepts like MenuHandle and createMaterialMenu from the public API.

This is similar to how Tooltips already work today (except that you get a controller to decide when to bring up the menu) and parallels the API we are designing for OverlayPortal and multi window.

gspencergoog commented:
> Curious to hear what you think...

This might make it easier to integrate, but it disables one of the use cases that we can currently allow: creating context menus that are placed with a global position (often from a pointer position, not a widget position), which is what the `create_material_menu.1.dart` example shows.

Also, unless you need the global positioning use case, we already have most of what you're talking about: you don't need to use `createMaterialMenu` or a `MenuAnchor` if you use `MenuBar` or `SubmenuButton` directly. The only time you need the lower level API is if you want to have a menu on something other than a plain button, or want to specify the opening location yourself.

However, we might be able to change the `MenuAnchor` so that it has an API closer to what you propose, because I do like that the menu description appears to be part of the widget tree in the same way as `MenuBar`. I worry a little that there will be ways in which it won't act the same, and thus be surprising, since it's actually drawn in the `Overlay`, at least until we get `OverlayPortal` (I'm mainly thinking of cases where they begin invalidating the widgets at the wrong time in the build).

If we eliminate `createMaterialMenu` and move its attributes to `MenuAnchor`, but still allow an optional global position to be specified by the `open()`, maybe that is sufficient?

I am a bit sad about not committing this yet...but I also don't want to remove API mere days after introducing it.

goderbauer commented:
> This might make it easier to integrate, but it disables one of the use cases that we can currently allow: creating context menus that are placed with a global position (often from a pointer position, not a widget position), which is what the create_material_menu.1.dart example shows.

For that my thinking was that you'd wrap the entire area in which you want the context menu to appear in the MenuAnchor and then the `.open` call could take an optional offset into the boundingbox defined by the MenuAnchor to position it based on a PointerEvent.

gspencergoog commented:
Michael and I talked about this, and I'm going to give it a try and see if I can refine the API surface to match his idea more closely.

gspencergoog commented:
Closing in favor of https://github.com/flutter/flutter/pull/112239 since GitHub is just getting too slow to be able to handle this PR with more comments.  Moving to a new PR in the hope that it'll be faster without the history, and also to preserve this iteration.
